### PR TITLE
[CELEBORN-83][FOLLOWUP] Fix various bugs when using HDFS as storage.

### PR DESCRIPTION
**Pull request by FMX**
# [CELEBORN-83][FOLLOWUP] Fix various bugs when using HDFS as storage.

### What changes were proposed in this pull request?
1.Fix class cast exception.
2.Fix clean hdfs expired app dir.
3.Fix npe when destroy file writers.
4.Adjust disk not usable log level.
5.Close hdfs output streams.
6.Change hdfs log level to reduce log size.
7.Add instructions for buffer size using HDFS.
8.Fix data loss when reading from HDFS.

### How was this patch tested?

/cc @related-reviewer

/assign @main-reviewer

### CODE REVIEW

File: client/src/main/java/org/apache/celeborn/client/ShuffleClient.java [hunk 1]
```diff
@@ -102,6 +102,8 @@ public static FileSystem getHdfsFs(CelebornConf conf) {
       synchronized (ShuffleClient.class) {
         if (null == hdfsFs) {
           Configuration hdfsConfiguration = new Configuration();
+          // enable fs cache to avoid too many fs instances
+          hdfsConfiguration.set("fs.hdfs.impl.disable.cache", "false");
```
Comment (pan3793) [inline client/src/main/java/org/apache/celeborn/client/ShuffleClient.java:106]: Emm... does Hadoop `Configuration` support `setIfMissing` semantic like `SparkConf`?

File: client/src/main/java/org/apache/celeborn/client/read/DfsPartitionReader.java [hunk 2]
```diff
@@ -98,30 +98,70 @@ public DfsPartitionReader(
...
                     byte[] buffer = new byte[(int) length];
-                    hdfsInputStream.readFully(offset, buffer);
-                    results.add(Unpooled.wrappedBuffer(buffer));
-                    currentChunkIndex.incrementAndGet();
+                    try {
+                      hdfsInputStream.readFully(offset, buffer);
+                    } catch (IOException e) {
+                      logger.warn(
+                          "read hdfs "
+                              + location.getStorageInfo().getFilePath()
```
Comment (pan3793) [inline client/src/main/java/org/apache/celeborn/client/read/DfsPartitionReader.java:125]: why not use `{}` as placeholder here?

Comment (FMX) [reply client/src/main/java/org/apache/celeborn/client/read/DfsPartitionReader.java]: Should use {}.

File: conf/log4j2.xml.template [hunk 3]
```diff
@@ -33,9 +33,13 @@
             <PatternLayout pattern="%d{yy/MM/dd HH:mm:ss,SSS} %p [%t] %c{1}: %m%n%ex"/>
         </Console>
     </Appenders>
+
     <Loggers>
         <Root level="INFO">
             <AppenderRef ref="stdout"/>
         </Root>
+        <Logger name="org.apache.hadoop.hdfs" level="WARN" additivity="false">
+                    <Appender-ref ref="stdout" level="WARN" />
```
Comment (pan3793) [inline conf/log4j2.xml.template:42]: could you please correct the indent?

File: worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/FileWriter.java [hunk 4]
```diff
@@ -105,7 +103,17 @@ public FileWriter(
     if (!fileInfo.isHdfs()) {
       channel = new FileOutputStream(fileInfo.getFilePath()).getChannel();
     } else {
-      stream = StorageManager.hdfsFs().create(fileInfo.getHdfsPath(), true);
+      try {
+        StorageManager.hdfsFs().create(fileInfo.getHdfsPath(), true).close();
```
Comment (pan3793) [inline worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/FileWriter.java:107]: `hdfsFs` is a little redundant name, suggest to rename it to `hadoopFs`, in all place, the reason is:

- we may support other hadoop compatible file system throw hadoop FileSystem API in the future, e.g. s3 oss alluxio
- in hdfsFs, "fs" is redundant

File: common/src/main/java/org/apache/celeborn/common/meta/FileInfo.java [hunk 5]
```diff
@@ -126,12 +125,16 @@ public UserIdentifier getUserIdentifier() {
...
-      hdfsFs.delete(getHdfsWriterSuccessPath(), false);
-      hdfsFs.delete(getHdfsIndexPath(), false);
-      hdfsFs.delete(getHdfsSortedPath(), false);
+      try {
+        hdfsFs.delete(getHdfsPath(), false);
+        hdfsFs.delete(getHdfsWriterSuccessPath(), false);
+        hdfsFs.delete(getHdfsIndexPath(), false);
+        hdfsFs.delete(getHdfsSortedPath(), false);
+      } catch (Exception e) {
+        // ignore delete exceptions because some other workers might be deleting the directory
```
Comment (pan3793) [inline common/src/main/java/org/apache/celeborn/common/meta/FileInfo.java:136]: can we logInfo the the e.getMessage at least?

**Comment by FMX** [comment]
Ping @pan3793. How about now?

File: client/src/main/java/org/apache/celeborn/client/ShuffleClient.java (hunk 1 above)
Comment (pan3793) [reply client/src/main/java/org/apache/celeborn/client/ShuffleClient.java:106]: I'm concerning this hardcode, if no other options, please at least add a info log here.

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/storage/Flusher.scala [hunk 6]
```diff
@@ -212,6 +212,7 @@ private[worker] class LocalFlusher(
   override def toString(): String = {
     s"LocalFlusher@$flusherId-$mountPoint"
   }
+
```
Comment (pan3793) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/storage/Flusher.scala:215]: please remove unnecessary changes in this file

File: client/src/main/java/org/apache/celeborn/client/ShuffleClient.java (hunk 1 above)
Comment (FMX) [reply client/src/main/java/org/apache/celeborn/client/ShuffleClient.java:109]: It is necessary to override cluster setting here.

Comment (FMX) [reply client/src/main/java/org/apache/celeborn/client/ShuffleClient.java:109]: Logs added.

**Review by pan3793 — APPROVED** [review]
